backend/Application/Aggregates/Contract/Entities/ContractSnapshot.cs

```diff
@@ -0,0 +1,198 @@
+                address,
+                latestContractSnapshot.ContractName,
+                moduleReferenceEvent?.ModuleReference ?? latestContractSnapshot.ModuleReference,
+                amount,
+                source
+            );
+
+            await repository.AddAsync(contractSnapshot);
+        }
+    }
+
+    private static async Task AddInitialContractSnapshots(
+        IEnumerable<(ulong ContractAddressIndex, ulong ContractAddressSubIndex)> addressesInitialized,
+        IList<ContractEvent> contractEventsAdded,
+        IContractRepository repository, 
+        IModuleReadonlyRepository moduleReadonlyRepository,
+        ImportSource source
+        )
+    {
+        foreach (var contractAddress in addressesInitialized)
+        {
+            var address = new ContractAddress(contractAddress.ContractAddressIndex, contractAddress.ContractAddressSubIndex);
+            var contractEvents = contractEventsAdded
+                .Where(ce => ce.ContractAddressIndex == contractAddress.ContractAddressIndex && ce.ContractAddressSubIndex == contractAddress.ContractAddressSubIndex)
+                .OrderBy(ce => ce.BlockHeight)
+                .ThenBy(ce => ce.TransactionIndex)
+                .ThenBy(ce => ce.EventIndex)
+                .ToList();
+
+            var contractInitializationEvent = (contractEvents.First().Event as ContractInitialized)!;
+            var amount = GetAmount(contractEvents, address, 0);
+            
+            var blockHeight = contractEvents.Last().BlockHeight;
+            
+            var moduleReferenceEvent = GetLatestModuleReferenceContractLinkEvent(address, moduleReadonlyRepository)!;
+
+            var contractSnapshot = new ContractSnapshot(
+                blockHeight,
+                address,
+                contractInitializationEvent.GetName(),
+                moduleReferenceEvent.ModuleReference,
+                amount,
+                source
+            );
+
+            await repository.AddAsync(contractSnapshot);
+        }
+    }
+
+    private static ModuleReferenceContractLinkEvent? GetLatestModuleReferenceContractLinkEvent(ContractAddress contractAddress, IModuleReadonlyRepository moduleReadonlyRepository) =>
+        moduleReadonlyRepository.GetModuleReferenceContractLinkEventInTransaction(contractAddress)
+            .OrderByDescending(m => m.BlockHeight)
+            .ThenByDescending(m => m.TransactionIndex)
+            .ThenByDescending(m => m.EventIndex)
+            .FirstOrDefault();
+
+    internal static double GetAmount(
+        IEnumerable<ContractEvent> contractEvents,
+        ContractAddress contractAddress,
+        double initialAmount)
+    {
+        var amount = initialAmount;
+        foreach (var contractEvent in contractEvents)
+        {
+            switch (contractEvent.Event)
+            {
+                case ContractInitialized contractInitialized:
+                    amount += (long)contractInitialized.Amount;
```

Comment:
**Issue:** In this function we are casting from unsigned (`ulong`) to signed integer (`long`) and then to implicitly from `long` to `double`, all of these conversions can cause data loss.